### PR TITLE
Implement ViewClusterRefresher caching synchronization

### PR DESCRIPTION
**Pull request by qqu0127**
### Issues

- [X] My PR addresses the following Helix issues and references them in the PR description:
Resolves https://github.com/apache/helix/issues/2198 

### Description

- [X] Here are some details about my PR, including screenshots of any UI changes:

A few terms used:
1. Source clusters -- the source clusters normally from different data centers that the view aggregator aggregate data from
2. View cluster data -- the aggregated cluster data view from the source clusters
3. Local cache -- the local copy of view cluster data

This PR Implements a cache synchronization mechanism to refresh local cache in `ViewClusterRefresher` from view cluster data. This is to provide extra resilience in case of data change externally in view cluster data. If that happens, the previous implementation will lead to eventual data inconsistency.

The cache is refactored as in `_viewClusterDataCache` to be a snapshot of remote view cluster data, the cache refresh happens at the beginning of each periodic refresh event. A data refresh from source cluster is triggered if a discrepancy is detected between local cache and remote view cluster data (source of truth)

### Tests

- [X] The following tests are written for this issue:

New test case in `TestHelixViewAggregator` and `TestBasicClusterDataCache`

- The following is the result of the "mvn test" command on the appropriate module:

[INFO] 
[INFO] --- maven-bundle-plugin:5.1.4:install (default-install) @ helix-view-aggregator ---
[INFO] Installing org/apache/helix/helix-view-aggregator/1.0.5-SNAPSHOT/helix-view-aggregator-1.0.5-SNAPSHOT.jar
[INFO] Writing OBR metadata
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for Apache Helix 1.0.5-SNAPSHOT:
[INFO] 
[INFO] Apache Helix ....................................... SUCCESS [  1.428 s]
[INFO] Apache Helix :: Metrics Common ..................... SUCCESS [  2.253 s]
[INFO] Apache Helix :: Metadata Store Directory Common .... SUCCESS [  1.068 s]
[INFO] Apache Helix :: ZooKeeper API ...................... SUCCESS [  2.445 s]
[INFO] Apache Helix :: Helix Common ....................... SUCCESS [  0.877 s]
[INFO] Apache Helix :: Core ............................... SUCCESS [ 10.478 s]
[INFO] Apache Helix :: Admin Webapp ....................... SUCCESS [  1.547 s]
[INFO] Apache Helix :: Restful Interface .................. SUCCESS [  3.015 s]
[INFO] Apache Helix :: Distributed Lock ................... SUCCESS [  0.714 s]
[INFO] Apache Helix :: HelixAgent ......................... SUCCESS [  1.369 s]
[INFO] Apache Helix :: Recipes ............................ SUCCESS [  0.020 s]
[INFO] Apache Helix :: Recipes :: Rabbitmq Consumer Group . SUCCESS [  1.135 s]
[INFO] Apache Helix :: Recipes :: Rsync Replicated File Store SUCCESS [  0.956 s]
[INFO] Apache Helix :: Recipes :: distributed lock manager  SUCCESS [  0.771 s]
[INFO] Apache Helix :: Recipes :: distributed task execution SUCCESS [  0.754 s]
[INFO] Apache Helix :: Recipes :: service discovery ....... SUCCESS [  0.733 s]
[INFO] Apache Helix :: View Aggregator .................... SUCCESS [  1.220 s]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  31.472 s
[INFO] Finished at: 2022-08-18T11:08:10-07:00
[INFO] ------------------------------------------------------------------------

### Changes that Break Backward Compatibility (Optional)

- My PR contains changes that break backward compatibility or previous assumptions for certain methods or API. They include:

(Consider including all behavior changes for public methods or API. Also include these changes in merge description so that other developers are aware of these changes. This allows them to make relevant code changes in feature branches accounting for the new method/API behavior.)

### Documentation (Optional)

- In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Commits

- My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

**Comment by xyuanlu** [comment]
Thanks for making the change. 
The description is pretty in detail. Could you please add one more minor thing that explains what does "fabric" mean here? It may help generic user to understand.
Also a dumb question. If the data change externally in view cluster data during refresh, will there be another callback and we will reach eventually consistency?

File: helix-core/src/main/java/org/apache/helix/common/caches/BasicClusterDataCache.java [hunk 1]
```diff
@@ -217,6 +218,23 @@ public void requireFullRefresh() {
...
+   * @return true if there is a change to local cache.
+   */
+  public boolean updateCache(HelixDataAccessor dataAccessor) {
+    Map<String, LiveInstance> liveInstance = new HashMap<>(getLiveInstances());
+    Map<String, ExternalView> externalView = new HashMap<>(getExternalViews());
+    Map<String, InstanceConfig> instanceConfig = new HashMap<>(getInstanceConfigMap());
+    requireFullRefresh();
+    refresh(dataAccessor);
+
+    return !(liveInstance.equals(getLiveInstances())
```
Comment (desaikomal) [inline helix-core/src/main/java/org/apache/helix/common/caches/BasicClusterDataCache.java:233]: i don't get this. you assigned liveInstace = getLiveInstances() on line: 227. and on line: 233 - you are again checking - which is just few cpu cycles after the first call

File: helix-view-aggregator/src/main/java/org/apache/helix/view/aggregator/HelixViewAggregator.java [hunk 2]
```diff
@@ -75,7 +75,7 @@ public HelixViewAggregator(String viewClusterName, String zkAddr) {
     _viewClusterManager = HelixManagerFactory
         .getZKHelixManager(_viewClusterName, generateHelixManagerInstanceName(_viewClusterName),
             InstanceType.SPECTATOR, zkAddr);
-    _refreshViewCluster = new AtomicBoolean(false);
+    _refreshViewCluster = new AtomicBoolean(true);
```
Comment (desaikomal) [inline helix-view-aggregator/src/main/java/org/apache/helix/view/aggregator/HelixViewAggregator.java:78]: why change default from 'false' to 'true'

File: helix-view-aggregator/src/main/java/org/apache/helix/view/aggregator/HelixViewAggregator.java [hunk 3]
```diff
@@ -191,14 +191,19 @@ private void handleSourceClusterEvent(ClusterViewEvent event) {
         _refreshViewCluster.set(true);
         break;
       case PeriodicViewRefresh:
-        if (!_refreshViewCluster.get()) {
+        // refresh local view cluster data cache
+        boolean cacheChanged = _viewClusterRefresher.refreshViewClusterDataCache();
+        if (cacheChanged) {
+          logger.info("Detected change in view cluster data, trigger a refresh");
+        } else if (!_refreshViewCluster.get()) {
```
Comment (desaikomal) [inline helix-view-aggregator/src/main/java/org/apache/helix/view/aggregator/HelixViewAggregator.java:198]: is refreshViewCluster ever getting set to FALSE ??

File: helix-view-aggregator/src/main/java/org/apache/helix/view/aggregator/HelixViewAggregator.java [hunk 4]
```diff
@@ -191,14 +191,19 @@ private void handleSourceClusterEvent(ClusterViewEvent event) {
         _refreshViewCluster.set(true);
         break;
       case PeriodicViewRefresh:
-        if (!_refreshViewCluster.get()) {
+        // refresh local view cluster data cache
```
[the remark this answers is not in the document]
Comment (qqu0127) [reply helix-view-aggregator/src/main/java/org/apache/helix/view/aggregator/HelixViewAggregator.java:194]: Short answer yes, it's already there.
There are two types of "refresh" here. 
1. between view cluster and source clusters (across fabrics), 
2. the other is between local view cluster data cache and remote view cluster source of truth.
The first one is much heavier and is only triggered during periodic refresh if a source cluster change is detected through notification, or item 2 is diverged. 
In the second case is more like integrity check, the divergence is unexpected. It's checked during each refresh cycle, and its implementation compares znode version as in `PropertyCache`, and it's actually selectively update

File: helix-core/src/main/java/org/apache/helix/common/caches/BasicClusterDataCache.java (hunk 1 above)
Comment (qqu0127) [reply helix-core/src/main/java/org/apache/helix/common/caches/BasicClusterDataCache.java:233]: The first one is to get a snapshot, and compare with the one after a refresh.

Comment (desaikomal) [reply helix-core/src/main/java/org/apache/helix/common/caches/BasicClusterDataCache.java:233]: but snapshot is taken just few cpu cycles before on line-227

File: helix-view-aggregator/src/main/java/org/apache/helix/view/aggregator/HelixViewAggregator.java (hunk 3 above)
Comment (qqu0127) [reply helix-view-aggregator/src/main/java/org/apache/helix/view/aggregator/HelixViewAggregator.java:198]: Yes, https://github.com/apache/helix/blob/5a4cfc71a5bca3e5d87bafc627eb3b571deb162d/helix-view-aggregator/src/main/java/org/apache/helix/view/aggregator/HelixViewAggregator.java#L355

File: helix-view-aggregator/src/main/java/org/apache/helix/view/aggregator/HelixViewAggregator.java (hunk 2 above)
Comment (qqu0127) [reply helix-view-aggregator/src/main/java/org/apache/helix/view/aggregator/HelixViewAggregator.java:78]: I think it makes more sense to trigger a refresh after startup (in the upcoming periodic refresh)

File: helix-core/src/main/java/org/apache/helix/common/caches/BasicClusterDataCache.java (hunk 1 above)
Comment (qqu0127) [reply helix-core/src/main/java/org/apache/helix/common/caches/BasicClusterDataCache.java]: refresh() happen and update the cache data in the middle of two `getLiveInstances()`, the data may be changed.

File: helix-view-aggregator/src/main/java/org/apache/helix/view/aggregator/HelixViewAggregator.java [hunk 5]
```diff
@@ -191,14 +191,19 @@ private void handleSourceClusterEvent(ClusterViewEvent event) {
...
+        boolean cacheChanged = _viewClusterRefresher.refreshViewClusterDataCache();
```
Comment (desaikomal) [inline helix-view-aggregator/src/main/java/org/apache/helix/view/aggregator/HelixViewAggregator.java:195]: just curious, refreshViewCluster can be set using the above call directly, correct?

**Review by desaikomal — APPROVED** [review]
Looks really complete, just minor comment.

File: helix-view-aggregator/src/main/java/org/apache/helix/view/aggregator/HelixViewAggregator.java (hunk 5 above)
Comment (qqu0127) [reply helix-view-aggregator/src/main/java/org/apache/helix/view/aggregator/HelixViewAggregator.java:195]: yes, other types of incoming event will set it.

**Comment by qqu0127** [comment]
This PR is ready to merge, approved by @desaikomal 
Failed tests are unrelated to helix-view-aggregator module

Commit message:

Implement ViewClusterRefresher caching synchronization
Implement a refresh mechanism to update view cluster local cache from remote source of truth data in case of external data change.

**Review by junkaixue — APPROVED** [review]

**Pull request merged**